File: Maps3DSamples/ApiDemos/java-app/src/main/java/com/example/maps3djava/routes/RouteRepository.java
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.maps3djava.routes;

import com.google.android.gms.maps.model.LatLng;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

/**
 * Represents the decoded route and step waypoints payload for the Java Routes API sample.
 */
class RouteData {
    private final String encodedPolyline;
    private final List<LatLng> navPoints;

    public RouteData(String encodedPolyline, List<LatLng> navPoints) {
        this.encodedPolyline = encodedPolyline;
        this.navPoints = navPoints;
    }

    public String getEncodedPolyline() {
        return encodedPolyline;
    }

    public List<LatLng> getNavPoints() {
        return navPoints;
    }
}

/**
 * A data repository responsible for executing background network tasks to compute driving
 * directions using the Google Maps Routes API (v2) in Java.
 */
public class RouteRepository {

    /**
     * Returns a Callable to fetch the route in a background thread pool.
     *
     * @param apiKey The API key to authenticate client requests.
     * @param origin Starting coordinate.
     * @param destination Destination coordinate.
     * @return A Callable producing [RouteData].
     */
    public Callable<RouteData> fetchRouteCallable(String apiKey, LatLng origin, LatLng destination) {
        return () -> {
            URL url = new URL("https://routes.googleapis.com/directions/v2:computeRoutes");
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("POST");
            connection.setRequestProperty("Content-Type", "application/json");
            connection.setRequestProperty("X-Goog-Api-Key", apiKey);
            connection.setRequestProperty("X-Goog-FieldMask", "routes.polyline.encodedPolyline,routes.legs.steps.startLocation");
            connection.setDoOutput(true);

            // Structure the JSON request body manually using standard JSONObjects
            JSONObject requestBody = new JSONObject();
            
            JSONObject originLatLng = new JSONObject()
                    .put("latitude", origin.latitude)
                    .put("longitude", origin.longitude);
            requestBody.put("origin", new JSONObject()
                    .put("location", new JSONObject().put("latLng", originLatLng)));

            JSONObject destLatLng = new JSONObject()
                    .put("latitude", destination.latitude)
                    .put("longitude", destination.longitude);
            requestBody.put("destination", new JSONObject()
                    .put("location", new JSONObject().put("latLng", destLatLng)));

            requestBody.put("travelMode", "DRIVE");

            // Stream payload to connection output
            try (OutputStreamWriter writer = new OutputStreamWriter(connection.getOutputStream())) {
                writer.write(requestBody.toString());
                writer.flush();
            }

            int responseCode = connection.getResponseCode();
            if (responseCode == HttpURLConnection.HTTP_OK) {
                StringBuilder response = new StringBuilder();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        response.append(line);
                    }
                }

                JSONObject jsonResponse = new JSONObject(response.toString());
                JSONArray routes = jsonResponse.getJSONArray("routes");
                if (routes.length() > 0) {
                    JSONObject route = routes.getJSONObject(0);
                    JSONObject polyline = route.getJSONObject("polyline");
                    String encodedPolyline = polyline.getString("encodedPolyline");

                    List<LatLng> navPoints = new ArrayList<>();
                    JSONArray legs = route.optJSONArray("legs");
                    if (legs != null && legs.length() > 0) {
                        JSONObject leg = legs.getJSONObject(0);
                        JSONArray steps = leg.optJSONArray("steps");
                        if (steps != null) {
                            for (int i = 0; i < steps.length(); i++) {
                                JSONObject step = steps.getJSONObject(i);
                                JSONObject startLocation = step.optJSONObject("startLocation");
                                if (startLocation != null) {
                                    JSONObject latLngObj = startLocation.getJSONObject("latLng");
                                    navPoints.add(new LatLng(
                                            latLngObj.getDouble("latitude"),
                                            latLngObj.getDouble("longitude")
                                    ));
                                }
                            }
                        }
                    }
                    navPoints.add(destination); // Cap route off with final destination
                    return new RouteData(encodedPolyline, navPoints);
                } else {
                    throw new Exception("No route details returned from the server.");
                }
            } else {
                StringBuilder errorResponse = new StringBuilder();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getErrorStream()))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        errorResponse.append(line);
                    }
                }
                throw new Exception("HTTP error " + responseCode + ": " + errorResponse);
            }
        };
    }
}

File: Maps3DSamples/ComposeDemos/app/src/main/java/com/example/composedemos/routes/RouteRepository.kt
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.composedemos.routes

import com.google.android.gms.maps.model.LatLng
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

data class RouteData(
    val encodedPolyline: String,
    val navPoints: List<LatLng>,
)

class RouteRepository {

    suspend fun fetchRoute(
        apiKey: String,
        origin: LatLng,
        dest: LatLng,
    ): RouteData = withContext(Dispatchers.IO) {
        val url = URL("https://routes.googleapis.com/directions/v2:computeRoutes")
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        connection.setRequestProperty("Content-Type", "application/json")
        connection.setRequestProperty("X-Goog-Api-Key", apiKey)
        connection.setRequestProperty("X-Goog-FieldMask", "routes.polyline.encodedPolyline,routes.legs.steps.startLocation")
        connection.doOutput = true

        val requestBody = JSONObject().apply {
            put(
                "origin",
                JSONObject().put(
                    "location",
                    JSONObject().put(
                        "latLng",
                        JSONObject().apply {
                            put("latitude", origin.latitude)
                            put("longitude", origin.longitude)
                        },
                    ),
                ),
            )
            put(
                "destination",
                JSONObject().put(
                    "location",
                    JSONObject().put(
                        "latLng",
                        JSONObject().apply {
                            put("latitude", dest.latitude)
                            put("longitude", dest.longitude)
                        },
                    ),
                ),
            )
            put("travelMode", "DRIVE")
        }

        OutputStreamWriter(connection.outputStream).use {
            it.write(requestBody.toString())
            it.flush()
        }

        val responseCode = connection.responseCode
        if (responseCode == HttpURLConnection.HTTP_OK) {
            val reader = BufferedReader(InputStreamReader(connection.inputStream))
            val response = StringBuilder()
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                response.append(line)
            }
            reader.close()

            val jsonResponse = JSONObject(response.toString())
            val routes = jsonResponse.getJSONArray("routes")
            if (routes.length() > 0) {
                val route = routes.getJSONObject(0)
                val polyline = route.getJSONObject("polyline")
                val encodedPolyline = polyline.getString("encodedPolyline")

                val navPoints = mutableListOf<LatLng>()
                val legs = route.optJSONArray("legs")
                if (legs != null && legs.length() > 0) {
                    val leg = legs.getJSONObject(0)
                    val steps = leg.optJSONArray("steps")
                    if (steps != null) {
                        for (i in 0 until steps.length()) {
                            val step = steps.getJSONObject(i)
                            val startLocation = step.optJSONObject("startLocation")
                            if (startLocation != null) {
                                val latLng = startLocation.getJSONObject("latLng")
                                navPoints.add(
                                    LatLng(
                                        latLng.getDouble("latitude"),
                                        latLng.getDouble("longitude"),
                                    ),
                                )
                            }
                        }
                    }
                }
                navPoints.add(dest) // Add destination as last point

                RouteData(encodedPolyline, navPoints)
            } else {
                throw Exception("No route returned from the Maps API.")
            }
        } else {
            val reader = BufferedReader(InputStreamReader(connection.errorStream))
            val response = StringBuilder()
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                response.append(line)
            }
            reader.close()
            throw Exception("API Error (HTTP $responseCode): $response")
        }
    }
}
